### PR TITLE
virtio-mmio: hypervisorless mode

### PR DESCRIPTION
**Pull request by danmilea**
OpenAMP library support for VIRTIO MMIO network, entropy and console in hypervisorless VIRTIO operation mode.

Validated using Zephyr environment based on https://github.com/OpenAMP/openamp-zephyr-staging/tree/virtio-exp.
Build tools for complete hypervisorless virtio system (Xilinx ZCU 102) are available here: https://github.com/danmilea/hypervisorless_virtio_zcu102/

Additional hypervisorless virtio information:
https://github.com/danmilea/hypervisorless_virtio_zcu102/blob/main/README_hypervisorless_virtio.md

### CODE REVIEW

File: lib/virtio_mmio/virtio_mmio.c
Comment (arnopo) [inline lib/virtio_mmio/virtio_mmio.c:40]: This one already exist `metal_sleep_usec`.

File: lib/virtio_mmio/virtio_mmio.c
Comment (arnopo) [inline lib/virtio_mmio/virtio_mmio.c:38]: Libmetal seems a good candidate to implement such service

File: lib/virtio_mmio/virtio_mmio.c
Comment (arnopo) [inline lib/virtio_mmio/virtio_mmio.c:378]: what contain this shared memory ? the bounce buffers?

File: lib/virtio_mmio/virtio_mmio.c
Comment (arnopo) [inline lib/virtio_mmio/virtio_mmio.c:36]: This seemsto be transformed to a generic callback that is called to send a signal to the guest or to the remote processor.
Similar feature is already implemented in virtio remoteproc:
https://github.com/OpenAMP/open-amp/blob/main/lib/remoteproc/remoteproc_virtio.c#L19

The callback should be provided by the system on init and stored in the the virtio_mmio_device struct

Comment (arnopo) [inline lib/virtio_mmio/virtio_mmio.c:36]: Using weak function in library is quite dangerous, we faced issue when we tried to use "weak" function in  libmetal. So I would prefer to avoid such mechanism

File: lib/virtio_mmio/virtio_mmio.c
Comment (arnopo) [inline lib/virtio_mmio/virtio_mmio.c:487]: Do you need to call virtio_mmio_hvl_ipi twice times? 
This one seems useless by looking at the code below

File: lib/virtio_mmio/virtio_mmio.c
Comment (arnopo) [inline lib/virtio_mmio/virtio_mmio.c:692]: I don't have a look yet to `virtio_mmio_hvl_add_bounce_buf` and `virtio_mmio_hvl_get_bounce_buf` functions
First, I need to understand the associated use case

File: lib/include/openamp/virtio_mmio.h
Comment (arnopo) [inline lib/include/openamp/virtio_mmio.h:428]: What about creating a virtio_mmio_hvl.c to split generic virtio_mmio and hypervisorless services?

**Review by arnopo — COMMENTED** [review]
Hi @danmilea 
Take time to review this PR.
From my point of view the main point here is the possibility to create a separate file such as `virtio_mmio_hvl.c ` separate the HVL specificities from the generic virtio mmio.
Do you think it is possible?

File: lib/include/openamp/virtio_mmio.h
Comment (danmilea) [reply lib/include/openamp/virtio_mmio.h:172]: The vrings may belong to multiple devices and have various sizes, so keeping track of multiple statically specified addresses can quickly become cumbersome. The current solution allows placing all of them in a specific memory area using linker scripts.

File: lib/virtio_mmio/virtio_mmio.c
Comment (danmilea) [reply lib/virtio_mmio/virtio_mmio.c:50]: Due to the variable vring size and dynamic placement, the memory pool for shared memory buffers is ultimately defined by __shmem0_pool_start and __shmem0_end. It would be possible to introduce a metal_device_io_region to act as indirection layer, but I don't see any immediate benefit when doing so.

File: lib/include/openamp/virtio_mmio.h
Comment (danmilea) [reply lib/include/openamp/virtio_mmio.h:246]: Virtio device drivers (aka virtio front-ends) have the liberty to allocate memory anywhere in their memory space. In order to be able to reuse these drivers without changes, a bounce buffer indirection layer is used. However, if the virtio front-end uses data buffers allocated from the shared memory region, "bouncing" is disabled. See /* check if buffer is already in pre-shared shared memory */ in virtio_mmio_hvl_add_bounce_buf in virtio_mmio.c

File: lib/include/openamp/virtio_mmio.h
Comment (danmilea) [reply lib/include/openamp/virtio_mmio.h:261]: Although not currently in use, yes, this would be the purpose of the irq field.

File: lib/virtio_mmio/virtio_mmio.c
Comment (danmilea) [reply lib/virtio_mmio/virtio_mmio.c:378]: Virtio device drivers (aka virtio front-ends) have the liberty to allocate memory anywhere in their memory space. In order to be able to reuse these drivers without changes, a bounce buffer indirection layer is used. However, if the virtio front-end uses data buffers allocated from the shared memory region, "bouncing" is disabled.

File: lib/virtio_mmio/virtio_mmio.c
Comment (danmilea) [reply lib/virtio_mmio/virtio_mmio.c:36]: Please share some information on the type of issues you encountered. In this case I think weak symbols provide a more readable implementation. Replacing them with function pointers is the other available option.

File: lib/virtio_mmio/virtio_mmio.c
Comment (danmilea) [reply lib/virtio_mmio/virtio_mmio.c:692]: These routines are the logical equivalents of add and get virtqueue buffers, but they operate on the shared memory buffer pool and handle the memcpy from and to the data buffers used by the virtio front-ends.

File: lib/virtio_mmio/virtio_mmio.c
Comment (danmilea) [reply lib/virtio_mmio/virtio_mmio.c:40]: The Zephyr implementation of metal_sleep_usec uses k_sleep. Since virtio device setup happens in the context of the kernel initialization routine, k_busy_wait needs to be used, which is what virtio_mmio_hvl_wait uses in the Zephyr reference implementation of hypervisorless virtio.

File: cmake/options.cmake
Comment (danmilea) [reply cmake/options.cmake]: Done.

File: lib/virtio_mmio/virtio_mmio.c
Comment (danmilea) [reply lib/virtio_mmio/virtio_mmio.c]: Fixed.

File: lib/include/openamp/virtio_mmio.h
Comment (danmilea) [reply lib/include/openamp/virtio_mmio.h]: Done in the latest version of the PR.

**Review by arnopo — COMMENTED** [review]
Look to me that this implementation does not respect the the virtio specification.
@danmilea Do you confirm?

If yes:
we could work on another way to Implement the "bounced buffer to avoid this. 
For incoming webinar demo, as it is on a staging branch, I can integrate if no time to improve that. But in this case:
- please rename the commit "virtio-mmio: temporary hypervisor-less mode"
- Addh comment explaining that is a tempory patch for demo that need rework to respect the virtio spec.

File: lib/virtio_mmio/virtio_mmio.c
Comment (danmilea) [reply lib/virtio_mmio/virtio_mmio.c:347]: In virtio spec terminology the device is the back-end and the driver in the guest is the front-end. In this case, the front-end (i.e. the mmio driver in open-amp) reads the registers.

File: lib/virtio_mmio/virtio_mmio.c
Comment (danmilea) [reply lib/virtio_mmio/virtio_mmio.c:359]: The front-end implementation, which is what this is, is tightly coupled with the back-end implementation from the Physical Machine Monitor. It also predates the release of the virtio spec v1.2 by more than a year.

File: lib/virtio_mmio/virtio_mmio.c
Comment (danmilea) [reply lib/virtio_mmio/virtio_mmio.c:454]: For devices which use multiple virtqueues, the same registers are used to provide the information for all the queues. The workflow is to select the queue (VIRTIO_MMIO_QUEUE_SEL), followed by VIRTIO_MMIO_QUEUE_PFN to set the queue address and then wait for ACK from the PMM. This repeats for the number of virtqueues. DRIVER_OK follows at the end of this pseudo-negotiation sequence. Introducing a wait may work, although it will complicate ISR processing (to be considered for potential enhancement).

File: lib/virtio/virtqueue.c
Comment (danmilea) [reply lib/virtio/virtqueue.c:220]: Yes, the bounce buffer layer copies the data from the original queue buffers and back, if the buffers were not allocated from the shared memory region in the first place. Btw, the HVL implementation predates the release of virtio spec 1.2 by more than a year. Aligning to the shared memory device specification requires additional research on feasibility and scope of work..

File: lib/virtio/virtqueue.c
Comment (danmilea) [reply lib/virtio/virtqueue.c:572]: Please rephrase or share additional information. vq_ring_add_buffer is called by the driver (aka front-end, aka guest). The virtio implementation in lib openamp is the front-end.

**Review by arnopo — COMMENTED** [review]
New review based on the good understanding of the virtio driver/devices role you implement. 
Your implementation show that we have to treat some topic as a next step for the virtio:
- support of multi virtio version
- shared memory management and associated memory allocation/allocator
- callback mechanism vs weak functions

For this step; i think the main point is to add comments that explain the design choices to be able to have the inputs when we will come back later on it.

File: lib/virtio_mmio/virtio_mmio.c
Comment (danmilea) [reply lib/virtio_mmio/virtio_mmio.c:359]: The initial target is to support transitional devices which allow legacy drivers.

File: lib/virtio_mmio/virtio_mmio.c
Comment (danmilea) [reply lib/virtio_mmio/virtio_mmio.c:454]: Yes, VIRTIO_MMIO_HVL_CFG_ACK is used by the physical machine monitor (PMM) to signal it has processed the configuration request. It is specific to hypervisorless virtio.

File: lib/include/openamp/virtio_mmio_hvl.h
Comment (danmilea) [reply lib/include/openamp/virtio_mmio_hvl.h:20]: In hypervisor-based deployments it's easy to signal individual devices, which in turns leads to their ISRs being executed. In hypervirsorless virtio deployments, this is replaced with a form of multiplexing by using a single source of notifications for all the devices. The callback framework allows individual device drivers to register their own callbacks if needed.

File: lib/virtio_mmio/virtio_mmio_hvl.c
Comment (danmilea) [reply lib/virtio_mmio/virtio_mmio_hvl.c:30]: I have introduced a new API call to set an IPI function pointer which is then used when HVL notifications are needed. For reasons of brevity, the implementation still makes use of weak symbols, but I mentioned in a comment the intent to replace them with non-OS specific equivalents (e.g. from libmetal).

File: lib/include/openamp/virtio_mmio_hvl.h
Comment (arnopo) [reply lib/include/openamp/virtio_mmio_hvl.h:28]: Please could you point me the zephyr code that uses this interface, to help me to understand the usage?

File: lib/virtio_mmio/virtio_mmio.c
Comment (arnopo) [reply lib/virtio_mmio/virtio_mmio.c:451]: Please you add a comment that specify that here it is a derogation of the virtio protocol for HVL

File: lib/virtio_mmio/virtio_mmio_hvl.c
Comment (arnopo) [reply lib/virtio_mmio/virtio_mmio_hvl.c:30]: I did not find the comment? is is comment in the review or comment in the code?
Having a comment in code such as a TODO would be nice.

**Review by arnopo — COMMENTED** [review]
I would like to see more comments in the code (see my revision comments) so that we don't have to refer to this PR in 6 months, but have the key information directly in the code. With that, for the virtio-exp branch, I guess we can merge.

File: lib/include/openamp/virtio_mmio_hvl.h
Comment (danmilea) [reply lib/include/openamp/virtio_mmio_hvl.h:28]: virtio_mmio_hvl_cb_set is used in virtio_mmio_device_init (virtio_mmio.c)  to set callbacks for each device which operates in hypervisorless mode. Currently the callback is virtio_mmio_isr, but the parameter is device dependent.

File: lib/virtio_mmio/virtio_mmio_hvl.c
Comment (danmilea) [reply lib/virtio_mmio/virtio_mmio_hvl.c:30]: The comment is right above the weak symbol declarations. I have reworded the comment and introduced a TODO item.

**Review by arnopo — COMMENTED** [review]
Some topics ( callbacks, memories management virtio spec version) will will probably need to been re discussed/improved before merging on main, but  let's move by steps.

**Review by arnopo — APPROVED** [review]

**Pull request merged**
